File: packages/agentspec-core/src/linter.ts
/**
 * Linter module — static analysis of AgentSpec files
 */

import type { AgentSpecFile, LintIssue } from './types';

function isValidISODate(str: string): boolean {
  const date = new Date(str);
  return !isNaN(date.getTime());
}

export function lintSpec(spec: AgentSpecFile): { valid: boolean; errors: LintIssue[]; warnings: LintIssue[]; suggestions: LintIssue[]; score: number } {
  const errors: LintIssue[] = [];
  const warnings: LintIssue[] = [];
  const suggestions: LintIssue[] = [];

  // Check version
  if (!spec.version) {
    errors.push({
      code: 'MISSING_VERSION',
      message: 'Spec is missing a version field',
      severity: 'error',
      location: 'version',
    });
  }

  // Check agent
  if (!spec.agent) {
    errors.push({
      code: 'MISSING_AGENT',
      message: 'Spec is missing an agent field',
      severity: 'error',
      location: 'agent',
    });
  } else {
    // Check timestamps
    if (spec.agent.createdAt && !isValidISODate(spec.agent.createdAt)) {
      errors.push({
        code: 'INVALID_DATE_FORMAT',
        message: 'Agent createdAt has invalid date format',
        severity: 'error',
        location: 'agent.createdAt',
      });
    }
    if (spec.agent.updatedAt && !isValidISODate(spec.agent.updatedAt)) {
      errors.push({
        code: 'INVALID_DATE_FORMAT',
        message: 'Agent updatedAt has invalid date format',
        severity: 'error',
        location: 'agent.updatedAt',
      });
    }
  }

  // Check capabilities
  if (!spec.capabilities || spec.capabilities.length === 0) {
    errors.push({
      code: 'NO_CAPABILITIES',
      message: 'Spec has no capabilities defined',
      severity: 'error',
      location: 'capabilities',
    });
  }

  // Check boundaries
  if (!spec.boundaries || spec.boundaries.length === 0) {
    errors.push({
      code: 'NO_BOUNDARIES',
      message: 'Spec has no boundaries defined',
      severity: 'error',
      location: 'boundaries',
    });
  }

  // Check duplicate IDs within capabilities
  if (spec.capabilities) {
    const capIds = new Set<string>();
    for (const cap of spec.capabilities) {
      if (capIds.has(cap.id)) {
        errors.push({
          code: 'DUPLICATE_ID',
          message: `Duplicate capability ID: ${cap.id}`,
          severity: 'error',
          location: `capabilities.${cap.id}`,
        });
      }
      capIds.add(cap.id);
    }
  }

  // Check duplicate IDs within boundaries
  if (spec.boundaries) {
    const boundIds = new Set<string>();
    for (const bound of spec.boundaries) {
      if (boundIds.has(bound.id)) {
        errors.push({
          code: 'DUPLICATE_ID',
          message: `Duplicate boundary ID: ${bound.id}`,
          severity: 'error',
          location: `boundaries.${bound.id}`,
        });
      }
      boundIds.add(bound.id);
    }
  }

  // Check high-risk capabilities without boundaries
  if (spec.capabilities) {
    for (const cap of spec.capabilities) {
      if (cap.riskLevel === 'critical' || cap.riskLevel === 'high') {
        const hasBoundary = spec.boundaries?.some(
          (b) => b.type === cap.category
        );
        if (!hasBoundary) {
          errors.push({
            code: 'HIGH_RISK_NO_BOUNDARY',
            message: `High-risk capability "${cap.id}" has no corresponding boundary`,
            severity: 'error',
            location: `capabilities.${cap.id}`,
          });
        }
      }
    }
  }

  const score = Math.max(0, 100 - errors.length * 15 - warnings.length * 5 - suggestions.length * 2);

  return { valid: errors.length === 0, errors, warnings, suggestions, score };
}

export function lintSecurity(spec: AgentSpecFile): LintIssue[] {
  const issues: LintIssue[] = [];

  // Check for critical capabilities
  if (spec.capabilities) {
    for (const cap of spec.capabilities) {
      if (cap.riskLevel === 'critical') {
        issues.push({
          code: 'CRITICAL_CAPABILITY',
          message: `Critical capability detected: ${cap.id}`,
          severity: 'warning',
          location: `capabilities.${cap.id}`,
        });
      }
    }
  }

  // Check audit config
  if (spec.audit && !spec.audit.enabled) {
    issues.push({
      code: 'AUDIT_DISABLED',
      message: 'Audit logging is disabled',
      severity: 'warning',
      location: 'audit',
    });
  }

  return issues;
}

export function lintPerformance(spec: AgentSpecFile): LintIssue[] {
  const issues: LintIssue[] = [];

  // Check too many boundaries
  if (spec.boundaries && spec.boundaries.length > 50) {
    issues.push({
      code: 'TOO_MANY_BOUNDARIES',
      message: `Spec has ${spec.boundaries.length} boundaries, which may impact performance`,
      severity: 'warning',
      location: 'boundaries',
    });
  }

  // Check too many capabilities
  if (spec.capabilities && spec.capabilities.length > 100) {
    issues.push({
      code: 'TOO_MANY_CAPABILITIES',
      message: `Spec has ${spec.capabilities.length} capabilities, which may impact performance`,
      severity: 'warning',
      location: 'capabilities',
    });
  }

  return issues;
}

export function lintDocumentation(spec: AgentSpecFile): LintIssue[] {
  const issues: LintIssue[] = [];

  // Check agent description
  if (spec.agent && !spec.agent.description) {
    issues.push({
      code: 'MISSING_DESCRIPTION',
      message: 'Agent is missing a description',
      severity: 'warning',
      location: 'agent.description',
    });
  }

  // Check capability descriptions
  if (spec.capabilities) {
    for (const cap of spec.capabilities) {
      if (!cap.description) {
        issues.push({
          code: 'MISSING_DESCRIPTION',
          message: `Capability "${cap.id}" is missing a description`,
          severity: 'warning',
          location: `capabilities.${cap.id}.description`,
        });
      }
    }
  }

  return issues;
}

File: packages/agentspec-py/agentspec/linter.py
"""AgentSpec Linter - Static analysis of AgentSpec files"""

from __future__ import annotations

from datetime import datetime
from typing import List

from .types import AgentSpecFile, LintIssue, LintResult, LintSeverity


def _is_valid_iso_date(s: str) -> bool:
    try:
        datetime.fromisoformat(s)
        return True
    except (ValueError, TypeError):
        return False


def lint_spec(spec: AgentSpecFile) -> LintResult:
    errors: List[LintIssue] = []
    warnings: List[LintIssue] = []
    suggestions: List[LintIssue] = []

    if not spec.version:
        errors.append(LintIssue(
            code="MISSING_VERSION",
            message="Spec is missing a version field",
            severity=LintSeverity.ERROR, location="version",
        ))

    if not spec.agent:
        errors.append(LintIssue(
            code="MISSING_AGENT",
            message="Spec is missing an agent field",
            severity=LintSeverity.ERROR, location="agent",
        ))
    else:
        if spec.agent.created_at and not _is_valid_iso_date(spec.agent.created_at):
            errors.append(LintIssue(
                code="INVALID_DATE_FORMAT",
                message="Agent createdAt has invalid date format",
                severity=LintSeverity.ERROR, location="agent.createdAt",
            ))
        if spec.agent.updated_at and not _is_valid_iso_date(spec.agent.updated_at):
            errors.append(LintIssue(
                code="INVALID_DATE_FORMAT",
                message="Agent updatedAt has invalid date format",
                severity=LintSeverity.ERROR, location="agent.updatedAt",
            ))

    if not spec.capabilities or len(spec.capabilities) == 0:
        errors.append(LintIssue(
            code="NO_CAPABILITIES",
            message="Spec has no capabilities defined",
            severity=LintSeverity.ERROR, location="capabilities",
        ))

    if not spec.boundaries or len(spec.boundaries) == 0:
        errors.append(LintIssue(
            code="NO_BOUNDARIES",
            message="Spec has no boundaries defined",
            severity=LintSeverity.ERROR, location="boundaries",
        ))

    # Duplicate capability IDs
    if spec.capabilities:
        cap_ids: set = set()
        for cap in spec.capabilities:
            if cap.id in cap_ids:
                errors.append(LintIssue(
                    code="DUPLICATE_ID",
                    message=f"Duplicate capability ID: {cap.id}",
                    severity=LintSeverity.ERROR,
                    location=f"capabilities.{cap.id}",
                ))
            cap_ids.add(cap.id)

    # Duplicate boundary IDs
    if spec.boundaries:
        bound_ids: set = set()
        for bound in spec.boundaries:
            if bound.id in bound_ids:
                errors.append(LintIssue(
                    code="DUPLICATE_ID",
                    message=f"Duplicate boundary ID: {bound.id}",
                    severity=LintSeverity.ERROR,
                    location=f"boundaries.{bound.id}",
                ))
            bound_ids.add(bound.id)

    # High-risk capabilities without corresponding boundary
    if spec.capabilities:
        for cap in spec.capabilities:
            risk = cap.risk_level
            risk_val = risk.value if hasattr(risk, "value") else str(risk)
            if risk_val in ("critical", "high"):
                has_boundary = any(
                    b.type == cap.category
                    for b in (spec.boundaries or [])
                )
                if not has_boundary:
                    errors.append(LintIssue(
                        code="HIGH_RISK_NO_BOUNDARY",
                        message=f'High-risk capability "{cap.id}" has no corresponding boundary',
                        severity=LintSeverity.ERROR,
                        location=f"capabilities.{cap.id}",
                    ))

    score = max(0, 100 - len(errors) * 15 - len(warnings) * 5 - len(suggestions) * 2)
    return LintResult(
        valid=len(errors) == 0,
        errors=errors, warnings=warnings, suggestions=suggestions,
        score=score,
    )


def lint_security(spec: AgentSpecFile) -> List[LintIssue]:
    issues: List[LintIssue] = []
    if spec.capabilities:
        for cap in spec.capabilities:
            risk = cap.risk_level
            risk_val = risk.value if hasattr(risk, "value") else str(risk)
            if risk_val == "critical":
                issues.append(LintIssue(
                    code="CRITICAL_CAPABILITY",
                    message=f"Critical capability detected: {cap.id}",
                    severity=LintSeverity.WARNING,
                    location=f"capabilities.{cap.id}",
                ))
    if spec.audit and not spec.audit.enabled:
        issues.append(LintIssue(
            code="AUDIT_DISABLED",
            message="Audit logging is disabled",
            severity=LintSeverity.WARNING, location="audit",
        ))
    return issues


def lint_performance(spec: AgentSpecFile) -> List[LintIssue]:
    issues: List[LintIssue] = []
    if spec.boundaries and len(spec.boundaries) > 50:
        issues.append(LintIssue(
            code="TOO_MANY_BOUNDARIES",
            message=f"Spec has {len(spec.boundaries)} boundaries, which may impact performance",
            severity=LintSeverity.WARNING, location="boundaries",
        ))
    if spec.capabilities and len(spec.capabilities) > 100:
        issues.append(LintIssue(
            code="TOO_MANY_CAPABILITIES",
            message=f"Spec has {len(spec.capabilities)} capabilities, which may impact performance",
            severity=LintSeverity.WARNING, location="capabilities",
        ))
    return issues


def lint_documentation(spec: AgentSpecFile) -> List[LintIssue]:
    issues: List[LintIssue] = []
    if spec.agent and not spec.agent.description:
        issues.append(LintIssue(
            code="MISSING_DESCRIPTION",
            message="Agent is missing a description",
            severity=LintSeverity.WARNING, location="agent.description",
        ))
    if spec.capabilities:
        for cap in spec.capabilities:
            if not cap.description:
                issues.append(LintIssue(
                    code="MISSING_DESCRIPTION",
                    message=f'Capability "{cap.id}" is missing a description',
                    severity=LintSeverity.WARNING,
                    location=f"capabilities.{cap.id}.description",
                ))
    return issues
